docs: contribution guide (#302)
samba-in-kubernetes/samba-operator: docs/contrib-guide.md

# Contribution Guide

1. [Prerequisites](#prerequisites)
2. [Build](#build)
3. [Check](#check)
4. [Testing](#testing)
5. [Pull-request](#pull-request)
6. [License](#license)

Thank you for your time and effort to help us improve samba-operator.
Here are a few steps to get started, with reference to some useful
resources.

## Prerequisites

Development effort takes place using Linux environment and requires at
minimum:

1. [Go 1.18](https://golang.org/dl/) installed
2. [GitHub](https://github.com/) account

> 3. Development tools: git and make

Reviewer: ```suggestion
Development tools: git, make, and podman or docker
```

> 4. Testing: [minikube](https://minikube.sigs.k8s.io)

Reviewer: ```suggestion
4. Testing: [minikube](https://minikube.sigs.k8s.io) (or equivalent)
```

Development collaboration takes place via github's samba-in-kuberbetus
([SINK](https://github.com/samba-in-kubernetes/)) pages and facilities.

The top-level [Makefile](../Makefile) is the entry point for various
build commands. Few utilities are required during the build process,
either for bootstrapping the project, building it or checking the
validity of the source code itself. If not present in your `$PATH`,
those tools may be installed locally to the project (under the
local `.bin` directory) using:

```sh
  $ make build-tools
  ...
  $ ls .bin/
```

## Build

The following section describes how to build and test samba-operator.
A more detailed description of this process may be found in the
[developer-notes](./developer-notes.md) document.

Building the samba-operator is straightforward action with `make build`.
Upon successful build, the output binary is written to `bin/manager`:

```sh
  $ make build
  ...
  $ ls bin/
  manager
```

> 

Reviewer: Let's also add something like:
```
Note that the manager binary alone is not typically executed in a Kubernetes cluster. For that we generally build a container image. To build an OCI container image including the manager binary, run:

` ` `sh
make image-build
` ` `

The name of this image is controlled by the IMG and TAG variables and may be customized to fit your needs.
```

Author: ACK

A developer may override some of the default Makefile settings via
the `devel.mk` file at the project's root directory:

```sh
  $ cp devel.mk.sample devel.mk
  $ vim devel.mk
```

## Check

Before submitting any pull request, a developer must ensure that

> his code satisfy the following minimum code-style requirements:

Reviewer: grammar: satisfy -> satisfies

Author: ACK

> 1. Go code conform to the project's standards:

Reviewer: grammar: conform -> conforms

```sh
  $ make check-format
  $ make check-golangci-lint
  $ make check-revive
  $ make check-gosec
```

2. YAML files conform to the project's standards:
```sh
  $ make check-yaml

> ```

Reviewer: Shouldn't we just request that `make check` be run as it bundles most of these into one command?

Author: Fair enough. Will fix.

3. Unit-tests pass without errors:
```sh
  $ make test
```

The above steps may also be executed using single make rule:
`make check test`. If any of those steps fail they must first be

> fixed before any other developer will review your patches.

Reviewer: Ah, here it is.  Maybe move this up top to tell people the simpler approach first and then mention how `make check` is a composite of other commands?

Author: ACK

Reviewer 2: > Ah, here it is. Maybe move this up top to tell people the simpler approach first and then mention how `make check` is a composite of other commands?

I suggest emphasizing tht the contributor is encouraged to perform those tests/checks on their local changes but that the CI will automatically run them and changes can only be added to the project if all those automated ci checks  pass

## Testing

> Before submitting any pull request, a developer should test his code
> on a kubernetes cluster using samba-operator built-in test suite,
> typically, using [minikube](https://minikube.sigs.k8s.io) and a
> private image:

Reviewer: Even running minikube can be a PITA. I would recommend but not require this, especially in a word where draft PRs can be created. So:
```suggestion
Ideally, before submitting any pull request, a developer will test any code changes locally.
Local clusters are simpler to debug and, if one has not had code merged to the project before,
it does not require approval from the project maintainers. A prerequisite to running the
samba-operator test suite is a running Kubernetes cluster. Typically this will be [minikube](https://minikube.sigs.k8s.io) but other Kubernetes implementations can suffice.
An example using minikube and a private container image follows:
```

Author: OK. Here I wanted also to emphasize to the naive developer that there is more than `make test` in this project. That said, I accept your suggestion for re-phrasing.

```sh
  $ export IMG=quay.io/my-quay-username/samba-operator:latest
  $ echo IMG=${IMG} >> devel.mk
  $ make image-build
  $ make container-push

>   ...

Reviewer: I'd make these ellipsis clear that they're not to be typed by either commenting them (prefix a `#`) or just turning them into full blown comments like:
```suggestion
# Start minikube test cluster...
```

  $ minikube start \
      --driver=kvm2 \
      --bootstrapper kubeadm \
      --disk-size 32g \
      --memory 8192 \
      --cpus 2 \
      --insecure-registry "10.0.0.0/24" \
      --nodes=3 -\
      --extra-disks=2 \
      --feature-gates=EphemeralContainers=true

>   ...

Reviewer: ```suggestion
# Ensure cluster has basic functionality
```

Author: ACK

  $ kubectl get nodes
  $ make deploy
  $ kuberctl get pods -n samba-operator-system

>   ...

Reviewer: ```suggestion
# Start an AD DC inside k8s to support domain logins in test suite
```

Author: ACK

  $ ./tests/test-deploy-ad-server.sh

>   $ ./tests/test.sh

Reviewer: ```suggestion
# Run the test suite
./tests/test.sh
```
